File: incre-tests/synduce/ptree/mul.f
Inductive Tree = leaf Unit | node {Int, Tree, Tree};
Inductive PTree = pleaf Unit | pnode {Int, PList}
     with PList = pnil Unit | pcons {PTree, PList};

repr = fix (
  \f: PTree -> Compress Tree. \pt: PTree.
  match pt with
    pleaf _ -> leaf unit
  | pnode {a, xs} ->
    let l2t = fix (
      \g: PList -> Compress Tree. \xs: PList.
      match xs with
        pnil _ -> leaf unit
      | pcons {h, t} -> node {0, f h, g t}
      end
    ) in
      node {a, leaf unit, l2t xs}
  end
);

spec = fix (
  \f: Tree -> Int. \t: Tree.
  match t with
    leaf _ -> 1
  | node {a, l, r} -> * a (* (f l) (f r))
  end
);

main = \x: PTree. spec (repr x);

/*l0 = pnil unit;
t1 = pnode {5, l0};
l1 = pcons {t1, pcons {t1, l0}};
t2 = pnode {3, l1};
repr t2;*/

File: ml-benchmark/synduce/ptree/mul.ml
type tree = Leaf of unit | Node of int * tree * tree
type 'a plist = PNil of unit | PCons of 'a * 'a plist
type ptree = PLeaf of unit | PNode of int * ptree plist

val repr: ptree -> tree compress
let rec repr pt =
  match pt with
  | PLeaf _ -> Leaf ()
  | PNode (a, xs) ->
      let rec l2t xs =
        match xs with
        | PNil _ -> Leaf ()
        | PCons (h, t) -> Node (0, repr h, l2t t)
      in
      Node (a, Leaf (), l2t xs)

let rec spec t =
  match t with
  | Leaf _ -> 1
  | Node (a, l, r) -> a * (spec l * spec r)

let program x = spec (repr x)